### ISSUE TITLE
listServices method throwing exception

### ISSUE THREAD

**Issue by markmclaren** [spotify/docker-client#689]
## Description

Whilst attempting to use the listServices method, an exception is produced

## How to reproduce

```
DockerClient docker = DefaultDockerClient.fromEnv().build();
List<Service> services = docker.listServices();
docker.close();
```

## What do you expect

I expected it to work without an exception.

## What happened instead

Exception thrown.

## Software:

- `docker version`: 1.13.1
- Spotify's docker-client version: 8.1.2

## Full backtrace

```
Exception in thread "main" com.spotify.docker.client.exceptions.DockerRequestException: Request error: GET https://blah.com:12345/services: 200
	at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:2165)
	at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:2094)
	at com.spotify.docker.client.DefaultDockerClient.listServices(DefaultDockerClient.java:1680)
	at edu.bristol.webfarmrepdockerclientexperiment.Test.main(Test.java:21)
Caused by: javax.ws.rs.client.ResponseProcessingException: com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.spotify.docker.client.messages.mount.Driver, problem: abstract types either need to be mapped to concrete types, have custom deserializer, or be instantiated with additional type information
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@582571ea; line: 1, column: 4164] (through reference chain: java.util.ArrayList[2]->com.spotify.docker.client.messages.swarm.Service["Spec"]->com.spotify.docker.client.messages.swarm.ServiceSpec["TaskTemplate"]->com.spotify.docker.client.messages.swarm.TaskSpec["ContainerSpec"]->com.spotify.docker.client.messages.swarm.ContainerSpec["Mounts"]->java.util.ArrayList[1]->com.spotify.docker.client.messages.mount.Mount["VolumeOptions"]->com.spotify.docker.client.messages.mount.VolumeOptions["DriverConfig"])
	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:869)
	at org.glassfish.jersey.client.JerseyInvocation.access$800(JerseyInvocation.java:92)
	at org.glassfish.jersey.client.JerseyInvocation$6.completed(JerseyInvocation.java:836)
	at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:198)
	at org.glassfish.jersey.client.ClientRuntime.access$300(ClientRuntime.java:79)
	at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:180)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:340)
	at org.glassfish.jersey.client.ClientRuntime$3.run(ClientRuntime.java:210)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.spotify.docker.client.messages.mount.Driver, problem: abstract types either need to be mapped to concrete types, have custom deserializer, or be instantiated with additional type information
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@582571ea; line: 1, column: 4164] (through reference chain: java.util.ArrayList[2]->com.spotify.docker.client.messages.swarm.Service["Spec"]->com.spotify.docker.client.messages.swarm.ServiceSpec["TaskTemplate"]->com.spotify.docker.client.messages.swarm.TaskSpec["ContainerSpec"]->com.spotify.docker.client.messages.swarm.ContainerSpec["Mounts"]->java.util.ArrayList[1]->com.spotify.docker.client.messages.mount.Mount["VolumeOptions"]->com.spotify.docker.client.messages.mount.VolumeOptions["DriverConfig"])
```

**Comment by davidxia** [comment]
Thanks for the report and PR. I'm unable to repro this, however, with your code on docker 1.13.1 and docker-client 8.1.2. Does `DefaultDockerClientTest.testListServices()` work for you? Do you have a service that you created that docker-client can't deserialize properly? What does `docker service ls` return?

**Comment by markmclaren** [comment]
I've updated the full back trace - I originally managed to remove some of the helpful parts of the exception.

I'm also struggling to build the 8.1.2 docker client source code on Windows/Java 8.  The CompressedDirectoryTest.testFileWithIgnore test fails saying "java.nio.file.InvalidPathException: Illegal char <*> at index 10: !subdir3\\*.keep".  I apologise as I appreciate that isn't very helpful.

I'm running my development code using the 8.1.2 jar binary against a remote Azure (Docker for Azure) hosted docker host.  Using "docker service ls" does appear to work fine.  

My guess from the exception is that the docker client doesn't deserialize the Driver section properly.  Using the Docker Engine API directly this looks to be typically:

```
"VolumeOptions" : {
    "DriverConfig" : {
        "Name" : "docker4x/cloudstor:azure-v1.13.1-ga-2"
    }
}
```
My fix was to make Driver.java look more like Mount.java which seemed to fix the issue I was seeing.

**Comment by johnflavin** [comment]
Thanks for that update. This makes more sense to me, and it looks like your PR #690 is exactly what is needed to fix the issue.

From my reading of your comments, I think to reproduce the issue you would need to create a service with some kind of option in `ContainerSpec.mounts().volumeOptions().driver()`, then attempt to read the service back with `listServices()` or `inspectService(id)`. There is no `@JsonCreator` for the `"DriverConfig"` key, so if anything were there the code would fail to deserialize in exactly the way you have described. But I don't really know what volume driver options are available, so I don't know how to create the service that would reproduce the issue.

Do you think you could produce a test case that can reproduce this issue and add it to #690?

**Comment by markmclaren** [comment]
I'm also not sure what valid values for name and option are (I'm a relative newcomer to Docker).  I settled for testing the deserialization in the manner of ImageInfoTest and ContainerInfoTest.  

Incidentally, it looked to me like there is problem with ImageInfoTest (setup is for ContainerInfo and test is against ImageInfo, should they both be ImageInfo?).

Weirdly in the Travis build, 4 builds passed and 2 builds failed.  The failures didn't look like anything to do with anything I had changed.

https://travis-ci.org/spotify/docker-client/builds/213573012
